### PR TITLE
Fix reentrancy vulnerability in example Crowfund.refund()

### PR DESCRIPTION
**Pull request by hitsuzen-eth**
Fix reentrancy vulnerability Crowfund.refund()

### CODE REVIEW

**Comment by hitsuzen-eth** [comment]
My fix is useless.
There is no possibility to steal fund with a reentrancy attack due to 2300 gas limit.
BUT if a smart contract is used as a participant it's possible to lock all funds in the Crowdfund due to REVERT on send call which freeze all fund.

I don't know how to fix that without implementing a pull model instead of push.

**Comment by fubuloubu** [comment]
> I don't know how to fix that without implementing a pull model instead of push.

Sounds like this should be the fix then!

**Comment by hitsuzen-eth** [comment]
I think it's ok now.

Crowdfund.refund() will only refund fund of the msg.sender and one participant can participate multiple time which increase his participation.

File: docs/vyper-by-example.rst [hunk 1]
```diff
@@ -275,14 +275,14 @@ the funds will be sent to the  beneficiary at the end of the campaign deadline.
 Participants will be refunded their respective contributions if the total
 funding does not reach its target goal.
 
-.. literalinclude:: ../examples/crowdfund.vy
+.. literalinclude:: ../examples/crowdfund/crowdfund.vy
```
Comment (fubuloubu) [inline docs/vyper-by-example.rst:278]: You know, I actually prefer all the examples to not be inside each of these folders, don't think it adds anything.

File: examples/crowdfund/crowdfund.vy [hunk 2]
```diff
@@ -42,22 +30,13 @@ def finalize():
...
-        if i >= self.nextFunderIndex:
-            self.refundIndex = self.nextFunderIndex
-            return
-
-        sender: address = self.funders[i].sender
-        value: uint256 = self.funders[i].value
-        self.funders[i] = empty(Funder)
-        send(sender, value)
+    value: uint256 = self.funders[msg.sender]
+    self.funders[msg.sender] = empty(uint256)
```
Comment (fubuloubu) [inline examples/crowdfund/crowdfund.vy:40]: Can probably just be 0
```suggestion
    self.funders[msg.sender] = 0
```

**Review by fubuloubu — APPROVED** [review]

**Review by fubuloubu — APPROVED** [review]

**Pull request merged**